sender/nats.go

```diff
@@ -0,0 +1,181 @@
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301  USA
+ */
+
+package sender
+
+import (
+	"crypto/tls"
+	"fmt"
+	"github.com/nats-io/nats.go"
+	"github.com/telenornms/skogul"
+	"github.com/telenornms/skogul/encoder"
+	"strings"
+	"sync"
+)
+
+var natsLog = skogul.Logger("sender", "nats")
+
+/*
+Nats sender. A small Nats, non-jetstream, publisher implementing all
+Authentication & Authorization features in the nats golang client.
+*/
+
+type Nats struct {
+	Servers       string `doc:"Comma separated list of nats URLs"`
+	Subject       string `doc:"Subject to publish messages on"`
+	SubjectAppend string `doc:"Append this Metadata field to subject"`
+	Name          string `doc:"Client name"`
+	Username      string `doc:"Client username"`
+	Password      string `doc:"Client password"`
+	TLSClientKey  string `doc:"TLS client key file path"`
+	TLSClientCert string `doc:"TLS client cert file path"`
+	TLSCACert     string `doc:"CA cert file path"`
+	UserCreds     string `doc:"Nats credentials file path"`
+	NKeyFile      string `doc:"Nats nkey file path"`
+	Insecure      bool   `doc:"TLS InsecureSkipVerify"`
+	Encoder       skogul.EncoderRef
+	conOpts       *[]nats.Option
+	natsCon       *nats.Conn
+	once          sync.Once
+	init_error    error
+}
+
+// Verify configuration
+func (n *Nats) Verify() error {
+	if n.Subject == "" {
+		return skogul.MissingArgument("Subject")
+	}
+	if n.Servers == "" {
+		return skogul.MissingArgument("Servers")
+	}
+
+	//User Credentials, use either.
+	if n.UserCreds != "" && n.NKeyFile != "" {
+		return fmt.Errorf("Please configure usercreds or nkeyfile.")
+	}
+
+	return nil
+}
+
+func (n *Nats) init() error {
+
+	if n.Encoder.Name == "" {
+		n.Encoder.E = encoder.JSON{}
+	}
+
+	if n.Name == "" {
+		n.Name = "skogul"
+	}
+	n.conOpts = &[]nats.Option{nats.Name(n.Name)}
+
+	if n.UserCreds != "" {
+		*n.conOpts = append(*n.conOpts, nats.UserCredentials(n.UserCreds))
+	}
+
+	//Plain text passwords
+	if n.Username != "" && n.Password != "" {
+		if n.TLSClientKey != "" {
+			natsLog.Warnf("Using plain text password over a non encrypted transport!")
+		}
+		*n.conOpts = append(*n.conOpts, nats.UserInfo(n.Username, n.Password))
+	}
+
+	//TLS authentication, Note: Fix selfsigned certificates.
```

Comment:
Is this "Note:" fixed?

Reply:
Yea, lingering comments since the first tests with TLS. Fixed with the "Insecure(SkipVerify)" in the configuration.